File: frontend/src/lib/types/expense.ts
// Types for the Expenses surface. Mirrors the JSON returned by the
// `/api/expenses` + `/api/expense-reports` endpoints (backend
// `ExpenseResponse` / `ExpenseReportResponse`). Money fields arrive as
// numbers (backend `float(...)`); date/datetime fields are ISO strings.

import type { BadgeTone } from '$lib/components/ui/Badge.svelte';
import type { MoneyAmount, MoneyString } from '$lib/utils/money';

export type ExpenseStatus = 'draft' | 'submitted' | 'approved' | 'rejected' | 'reimbursed';

// Every value the `expenses.status` column can hold — the full mirror of the
// backend `ExpenseStatus` enum (`backend/app/models/expense.py`). Kept COMPLETE
// on purpose: a row can still arrive carrying any of them (the demo seed
// `backend/scripts/seed_extras.py` writes `rejected` + `reimbursed`, and a
// long-lived tenant may hold rows written before a transition changed), so
// every value must still render a badge. This is NOT the filter-chip list —
// see `EXPENSE_FILTER_STATUSES`.
export const EXPENSE_STATUSES: ExpenseStatus[] = [
	'draft',
	'submitted',
	'approved',
	'rejected',
	'reimbursed'
];

// Statuses no backend transition ever stamps, so a filter chip for them is a
// control that can never return a row. Every writer of `Expense.status` in the
// backend:
//   - `draft`     — the column default on insert (`models/expense.py:229`), and
//                   `reject_report` (`api/expenses.py:1386`) which puts a
//                   rejected report's children BACK to draft.
//   - `submitted` — `submit_report` (`api/expenses.py:1230`).
//   - `approved`  — `approve_report` (`api/expenses.py:1341`).
// That is the complete list under `backend/app/`. Nothing writes `rejected`
// (report rejection returns children to `draft`) and nothing writes
// `reimbursed` (there is no reimbursement transition anywhere — no route even
// sets `ExpenseReportStatus.reimbursed`). Only the demo seed
// `backend/scripts/seed_extras.py` writes them.
//
// Delete an entry here the moment its writer lands — `reject_report` stamping
// `ExpenseStatus.rejected` on its children, or a new reimbursement endpoint
// stamping `ExpenseStatus.reimbursed`.
const UNREACHABLE_EXPENSE_STATUSES: ExpenseStatus[] = ['rejected', 'reimbursed'];

// The subset offered as *filter chips* — derived by EXCLUSION from the full
// mirror, so a genuinely new status added above joins the chips by default and
// only a deliberate, justified entry in UNREACHABLE_EXPENSE_STATUSES keeps one
// out. The excluded values still live in the union and in
// EXPENSE_STATUS_LABELS (a legacy / seeded row must still render its badge),
// and the page appends whatever status is *actively* filtered to the chip row,
// so an explicit `?status=reimbursed` is never an invisible filter.
export const EXPENSE_FILTER_STATUSES: ExpenseStatus[] = EXPENSE_STATUSES.filter(
	(s) => !UNREACHABLE_EXPENSE_STATUSES.includes(s)
);

export const EXPENSE_STATUS_LABELS: Record<ExpenseStatus, string> = {
	draft: 'Draft',
	submitted: 'Submitted',
	approved: 'Approved',
	rejected: 'Rejected',
	reimbursed: 'Reimbursed'
};

/**
 * Badge tone per expense status. Hoisted out of `ExpenseModal`, which is where
 * it was first written and where it left a note saying it belonged here once
 * the list page converted. It now has three callers — the modal, the Expenses
 * table and the report-detail line table — which is exactly the shape
 * `frontend/CLAUDE.md` § Badge asks for.
 *
 * Total record: a status added to `ExpenseStatus` is a compile error here
 * rather than an untinted pill.
 *
 * `reimbursed` takes the `erp` tone — the measured purple the list page's rule
 * spelled by hand, doing the job that tone does elsewhere: handed off
 * downstream. Green would collapse it into `approved`, and "someone approved
 * this" and "the money went back" are different answers to the only question
 * an employee asks of this pill.
 */
export const EXPENSE_STATUS_TONES: Record<ExpenseStatus, BadgeTone> = {
	draft: 'accent',
	submitted: 'warning',
	approved: 'success',
	rejected: 'danger',
	reimbursed: 'erp'
};

export type ExpensePaymentMethod = 'out_of_pocket' | 'corporate_card' | 'virtual_card';

export const EXPENSE_PAYMENT_METHODS: ExpensePaymentMethod[] = [
	'out_of_pocket',
	'corporate_card',
	'virtual_card'
];

export const EXPENSE_PAYMENT_METHOD_LABELS: Record<ExpensePaymentMethod, string> = {
	out_of_pocket: 'Out of pocket',
	corporate_card: 'Corporate card',
	virtual_card: 'Virtual card'
};

export type ExpenseReportStatus =
	| 'draft'
	| 'submitted'
	| 'pending_approval'
	| 'approved'
	| 'rejected'
	| 'reimbursed'
	| 'cancelled';

export const EXPENSE_REPORT_STATUSES: ExpenseReportStatus[] = [
	'draft',
	'submitted',
	'pending_approval',
	'approved',
	'rejected',
	'reimbursed',
	'cancelled'
];

export const EXPENSE_REPORT_STATUS_LABELS: Record<ExpenseReportStatus, string> = {
	draft: 'Draft',
	submitted: 'Submitted',
	pending_approval: 'Pending Approval',
	approved: 'Approved',
	rejected: 'Rejected',
	reimbursed: 'Reimbursed',
	cancelled: 'Cancelled'
};

/**
 * Badge tone per expense-REPORT status. Two callers on the same page — the
 * reports table and the report-detail header, which the e2e suite reads as
 * `.report-title-block .badge`.
 *
 * `submitted` and `pending_approval` share `warning` because the report is
 * waiting on someone in both, which is what the pill is for. They keep their
 * own labels (and their own filter chips), so the states stay distinguishable
 * in text — SC 1.4.1 — while the colour answers the only scannable question.
 *
 * `cancelled` is `neutral`, not a grey tint: a withdrawn report is the absence
 * of a signal rather than a weak one. That is the same call `/payments` makes
 * for a `draft` run, and the opposite of the one it makes for a `voided`
 * payment — money that moved and came back is an event; a report nobody
 * pursued is not.
 */
export const EXPENSE_REPORT_STATUS_TONES: Record<ExpenseReportStatus, BadgeTone> = {
	draft: 'accent',
	submitted: 'warning',
	pending_approval: 'warning',
	approved: 'success',
	rejected: 'danger',
	reimbursed: 'erp',
	cancelled: 'neutral'
};

// A single policy-engine finding stamped onto `Expense.policy_violations` by the
// WF3 backend engine (`evaluate_expense`). `code` is a stable machine key
// (`category_limit`, `receipt_required`, `preapproval_required`,
// `per_diem_exceeded`, `mileage_amount_mismatch`), `message` is the human
// string the badge tooltip renders.
export interface PolicyViolation {
	code: string;
	message: string;
	policy_id?: string;
	/** Exact decimal strings — money never round-trips as a JS number. */
	limit?: string;
	actual?: string;
	/**
	 * `mileage_amount_mismatch` only — the working behind `limit`, as exact
	 * strings. `miles` is a distance, not money; `rate` is per-mile in
	 * `currency`.
	 */
	miles?: string;
	rate?: string;
	/** Currency `limit` (and, when the comparison resolved, `actual`) is in. */
	currency?: string;
	/**
	 * Present only when the expense could not be expressed in `currency`, so the
	 * rule fell closed and was flagged without a comparison. `actual` is then the
	 * expense's face amount, in `expense_currency`.
	 */
	comparison?: 'unresolved';
	expense_currency?: string;
}

export interface Expense {
	id: string;
	report_id: string | null;
	expense_date: string;
	merchant: string | null;
	category: string | null;
	description: string | null;
	amount: MoneyAmount;
	currency: string;
	// Rate-locked expression of `amount` in the owning report's currency
	// (issue #157). Exact decimal STRINGS — never parse into a float for
	// arithmetic. null when the expense isn't attached to a report.
	converted_amount: string | null;
	converted_currency: string | null;
	converted_fx_rate: string | null;
	converted_fx_locked_at: string | null;
	gl_account_id: string | null;
	receipt_file_key: string | null;
	receipt_url: string | null;
	payment_method: string;
	card_transaction_id: string | null;
	policy_violations: PolicyViolation[] | null;
	status: string;
	reimbursable: boolean;
	mileage_miles: number | null;
	created_at: string;
	updated_at: string;
}

export interface ExpenseListResponse {
	items: Expense[];
	total: number;
	page: number;
	page_size: number;
}

/** One currency's slice of `GET /api/expenses/summary`. */
export interface ExpenseCurrencyTotal {
	currency: string;
	/** Exact decimal string — never parse into a float for arithmetic. */
	total: MoneyString;
	count: number;
}

/**
 * Whole-set rollup from `GET /api/expenses/summary`, over the SAME filters the
 * list ran with.
 *
 * The KPI row used to derive its figures from the loaded page, so "Period
 * total" and "Pending" described 20 rows while the "Expenses" card beside them
 * described every row. `by_currency` is grouped rather than summed — adding EUR
 * to USD produces a figure denominated in nothing (see
 * `$lib/utils/currencyGroups`).
 */
export interface ExpenseSummary {
	total: number;
	by_status: Record<string, number>;
	by_currency: ExpenseCurrencyTotal[];
}

export interface ExpenseReport {
	id: string;
	report_number: string;
	title: string | null;
	employee_user_id: string;
	status: string;
	submitted_at: string | null;
	approved_at: string | null;
	approved_by: string | null;
	total_amount: MoneyAmount;
	// Exact `total_amount` (in `currency`) as a decimal string.
	total_amount_exact: string;
	currency: string;
	// `total_amount` re-expressed in the org reporting currency at the rate
	// locked on submit — the figure the CFO threshold gate compares. null
	// before submit or when no rate was available (the gate then requires CFO).
	reporting_amount: string | null;
	reporting_currency: string | null;
	reporting_fx_rate: string | null;
	reporting_fx_locked_at: string | null;
	notes: string | null;
	expenses: Expense[];
	created_at: string;
	updated_at: string;
}

export interface ExpenseReportListResponse {
	items: ExpenseReport[];
	total: number;
	page: number;
	page_size: number;
}

// `GET /api/expense-reports/{id}/summary` — computed from the report's expenses.
export interface ExpenseSummaryBucket {
	category?: string | null;
	status?: string | null;
	total: MoneyAmount;
	// Exact `total` as a decimal string.
	total_exact: string;
	// Lines with no usable rate lock — EXCLUDED from `total`.
	unconverted_count: number;
	count: number;
}

// Per-currency split of a report's lines. `original_amount` is the face value
// in that currency; `report_amount` is its rate-locked contribution to the
// report total. Both exact decimal strings.
export interface ExpenseCurrencyBucket {
	currency: string;
	count: number;
	original_amount: string;
	report_amount: string;
	unconverted_count: number;
}

export interface ExpenseReportSummary {
	total: MoneyAmount;
	// Exact `total`, denominated in `currency` — the report's own currency.
	// Every figure here is converted at each line's LOCKED rate, never a naive
	// cross-currency sum (issue #157).
	total_exact: string;
	currency: string;
	count: number;
	// Non-zero means the displayed totals are partial: some lines have no
	// usable rate lock and were excluded (they also block submission).
	unconverted_count: number;
	by_category: ExpenseSummaryBucket[];
	by_status: ExpenseSummaryBucket[];
	by_currency: ExpenseCurrencyBucket[];
}

// Payload shapes for create / update (request side). Money goes out as a
// number — the backend coerces to Decimal. Optional fields default server-side.
export interface ExpenseCreate {
	/**
	 * Required, `YYYY-MM-DD`. The column is NOT NULL and both `ExpenseCreate`
	 * and `ExpenseUpdate` type it as a bare `date` server-side — a null here is
	 * a 422 on create and (before the schema was tightened) a 500 on PATCH.
	 * `updateExpense` takes `Partial<ExpenseCreate>`, so a PATCH may still omit
	 * the field entirely; it may never send it empty.
	 */
	expense_date: string;
	merchant: string | null;
	category: string | null;
	description: string | null;
	/** Request side — the exact decimal text typed, never a JSON number. */
	amount: MoneyString;
	currency: string;
	gl_account_id: string | null;
	payment_method: string;
	reimbursable: boolean;
	/**
	 * Distance driven. Backed by the policy engine's `mileage_rate`, which
	 * flags a claim that isn't `miles x rate` (`mileage_amount_mismatch`) —
	 * so leaving this null is what tells the backend the line isn't a trip.
	 */
	mileage_miles?: number | null;
	report_id?: string | null;
}

export interface ExpenseReportCreate {
	report_number: string;
	title: string | null;
	currency: string;
	notes: string | null;
}

// ============================ WF3: Policies ============================

export interface ExpensePolicy {
	id: string;
	name: string;
	active: boolean;
	category: string | null;
	/**
	 * Currency every money threshold on this policy is denominated in.
	 * null = "the org's reporting currency" (resolved server-side at evaluation
	 * time) — the unit a bare threshold number has always implicitly had.
	 */
	threshold_currency: string | null;
	per_diem_amount: MoneyAmount;
	per_diem_currency: string | null;
	// A money-PER-MILE rate, not an amount: the policy engine flags a claim
	// that isn't `miles x rate`, so this is a factor in that product rather
	// than a figure the UI renders as currency.
	mileage_rate: number | null;
	category_limit: MoneyAmount;
	requires_preapproval_above: MoneyAmount;
	requires_receipt_above: MoneyAmount;
	rules: unknown | null;
	created_at: string;
	updated_at: string;
}

export interface ExpensePolicyCreate {
	name: string;
	active: boolean;
	category: string | null;
	threshold_currency: string | null;
	/** Request side — the exact decimal text typed, never a JSON number. */
	category_limit: MoneyString | null;
	requires_receipt_above: MoneyString | null;
	requires_preapproval_above: MoneyString | null;
	per_diem_amount: MoneyString | null;
	mileage_rate: number | null;
}

// ========================= WF3: Pre-approvals =========================

export type ExpensePreapprovalStatus = 'pending' | 'approved' | 'rejected';

export const EXPENSE_PREAPPROVAL_STATUSES: ExpensePreapprovalStatus[] = [
	'pending',
	'approved',
	'rejected'
];

export const EXPENSE_PREAPPROVAL_STATUS_LABELS: Record<ExpensePreapprovalStatus, string> = {
	pending: 'Pending',
	approved: 'Approved',
	rejected: 'Rejected'
};

/** Badge tone per pre-approval status — a decision pending, taken, or refused. */
export const EXPENSE_PREAPPROVAL_STATUS_TONES: Record<ExpensePreapprovalStatus, BadgeTone> = {
	pending: 'warning',
	approved: 'success',
	rejected: 'danger'
};

export interface ExpensePreapproval {
	id: string;
	requester_user_id: string;
	title: string;
	estimated_amount: MoneyAmount;
	currency: string;
	category: string | null;
	justification: string | null;
	status: string;
	decided_by: string | null;
	decided_at: string | null;
	expense_report_id: string | null;
	created_at: string;
	updated_at: string;
}

export interface ExpensePreapprovalCreate {
	title: string;
	/** Request side — the exact decimal text typed, never a JSON number. */
	estimated_amount: MoneyString;
	currency: string;
	category: string | null;
	justification: string | null;
}

// ===================== WF4: Corporate-card transactions =====================

export type ReconciliationStatus = 'unmatched' | 'matched' | 'ignored';

export const RECONCILIATION_STATUSES: ReconciliationStatus[] = [
	'unmatched',
	'matched',
	'ignored'
];

export const RECONCILIATION_STATUS_LABELS: Record<ReconciliationStatus, string> = {
	unmatched: 'Unmatched',
	matched: 'Matched',
	ignored: 'Ignored'
};

/**
 * Badge tone per card-transaction reconciliation status.
 *
 * `unmatched` is `warning`, not `danger`: a charge nobody has coded yet is
 * work outstanding, not a failure — and this table has no failure state to
 * confuse it with. `ignored` is `neutral` (flat), the deliberate "no signal"
 * chip: someone has decided this line needs nothing, which is the absence of a
 * signal rather than a weak one.
 */
export const RECONCILIATION_STATUS_TONES: Record<ReconciliationStatus, BadgeTone> = {
	unmatched: 'warning',
	matched: 'success',
	ignored: 'neutral'
};

// Mirrors the backend `CorporateCardTransactionResponse`. Money fields arrive as
// numbers (backend `float(...)` on the `Numeric(15,2)` column); date fields are
// ISO date strings. PII: only `card_last_four` is ever surfaced — never a PAN.
export interface CorporateCardTransaction {
	id: string;
	card_ref: string | null;
	card_last_four: string | null;
	virtual_card_id: string | null;
	txn_date: string;
	posted_date: string | null;
	merchant: string | null;
	amount: MoneyAmount;
	currency: string;
	external_txn_id: string | null;
	matched_expense_id: string | null;
	reconciliation_status: string;
	import_batch: string | null;
	created_at: string;
	updated_at: string;
}

export interface CardTransactionListResponse {
	items: CorporateCardTransaction[];
	total: number;
	page: number;
	page_size: number;
}

// A ranked candidate expense for reconciliation (from
// GET /{id}/match-suggestions) — the Expense plus a match score.
export interface CardMatchSuggestion {
	expense: Expense;
	score: number;
}

// POST /import-csv returns the shared `ImportResult.to_dict()` — `imported` /
// `skipped` / `errors`.
export interface CardImportResult {
	imported: number;
	skipped: number;
	errors: { row: number; message: string }[];
}

// POST /sync-virtual-cards return shape — created vs already-imported.
export interface SyncVirtualCardsResult {
	created: number;
	skipped: number;
}

File: backend/app/models/expense.py
"""Expense management — out-of-pocket + corporate-card expenses, reports,
policies, pre-approvals, and corporate-card-transaction reconciliation.

Five tenant-scoped tables in one module (mirrors contract.py / supplier_chat.py
patterns): inline ``organization_id`` + ``EntityMixin`` on every business table,
StrEnum status types mapped to ``Enum(..., native_enum=False, length=N)`` String
columns, money as ``Numeric(15, 2)`` (never float). A circular FK between
``expenses.card_transaction_id`` and ``corporate_card_transactions.matched_expense_id``
is broken with ``use_alter=True`` on one side so ``metadata.create_all`` orders
the DDL itself. See ``backend/docs/expense-management.md``.
"""

import enum
import uuid
from datetime import date, datetime
from decimal import Decimal

from sqlalchemy import (
    Boolean,
    Date,
    DateTime,
    Enum,
    ForeignKey,
    Index,
    Numeric,
    String,
    Text,
    text,
)
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base, EntityMixin, TimestampMixin


class ExpenseReportStatus(enum.StrEnum):
    draft = "draft"
    submitted = "submitted"
    pending_approval = "pending_approval"
    approved = "approved"
    rejected = "rejected"
    reimbursed = "reimbursed"
    cancelled = "cancelled"


class ExpenseStatus(enum.StrEnum):
    draft = "draft"
    submitted = "submitted"
    approved = "approved"
    rejected = "rejected"
    reimbursed = "reimbursed"


class ExpensePaymentMethod(enum.StrEnum):
    out_of_pocket = "out_of_pocket"
    corporate_card = "corporate_card"
    virtual_card = "virtual_card"


class ReconciliationStatus(enum.StrEnum):
    unmatched = "unmatched"
    matched = "matched"
    ignored = "ignored"


class PreapprovalStatus(enum.StrEnum):
    pending = "pending"
    approved = "approved"
    rejected = "rejected"


class ExpenseReport(Base, EntityMixin, TimestampMixin):
    """A grouping of expenses an employee submits for approval + reimbursement."""

    __tablename__ = "expense_reports"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    report_number: Mapped[str] = mapped_column(String(50), nullable=False)
    title: Mapped[str | None] = mapped_column(String(255))
    # Submitting User (control-plane User id; no cross-DB FK — plain uuid + index).
    employee_user_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), index=True)
    status: Mapped[ExpenseReportStatus] = mapped_column(
        Enum(ExpenseReportStatus, native_enum=False, length=30),
        default=ExpenseReportStatus.draft,
        nullable=False,
    )
    submitted_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    approved_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    approved_by: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True))

    # --- Money (Numeric, never float) -------------------------------------
    total_amount: Mapped[Decimal] = mapped_column(
        Numeric(15, 2), default=Decimal("0"), nullable=False
    )
    currency: Mapped[str] = mapped_column(String(3), default="USD")

    # --- Locked conversion into the ORG REPORTING currency (issue #157) ----
    # ``total_amount`` is denominated in this report's own ``currency``; the CFO
    # threshold (settings.expense_approval.cfo_threshold) is a bare number in the
    # org's reporting currency. These four columns snapshot the total in that
    # currency at SUBMIT time so the gate compares like with like and can't be
    # dodged by filing in a foreign currency. Same shape/semantics as
    # ``invoices.reporting_*`` (migration 0025). NULL = not established; the gate
    # then fails CLOSED (CFO sign-off required). See
    # ``services/expense_currency.py``.
    reporting_currency: Mapped[str | None] = mapped_column(String(3))
    reporting_amount: Mapped[Decimal | None] = mapped_column(Numeric(15, 2))
    reporting_fx_rate: Mapped[Decimal | None] = mapped_column(Numeric(18, 8))
    reporting_fx_locked_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))

    notes: Mapped[str | None] = mapped_column(Text)

    organization_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), nullable=False, index=True
    )

    expenses: Mapped[list["Expense"]] = relationship(back_populates="report")


class CorporateCardTransaction(Base, EntityMixin, TimestampMixin):
    """A corporate / virtual card transaction feed row, reconciled to an expense."""

    __tablename__ = "corporate_card_transactions"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    card_ref: Mapped[str | None] = mapped_column(String(255))
    card_last_four: Mapped[str | None] = mapped_column(String(4))
    virtual_card_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("virtual_cards.id"), index=True
    )
    txn_date: Mapped[date] = mapped_column(Date, nullable=False)
    posted_date: Mapped[date | None] = mapped_column(Date)
    merchant: Mapped[str | None] = mapped_column(String(255))

    # --- Money (Numeric, never float) -------------------------------------
    amount: Mapped[Decimal] = mapped_column(Numeric(15, 2), nullable=False)
    currency: Mapped[str] = mapped_column(String(3), default="USD")

    # Provider transaction id — drives import idempotency (partial unique index).
    external_txn_id: Mapped[str | None] = mapped_column(String(255))
    # Plain FK leg of the expenses↔transactions cycle (the alter side is on
    # Expense.card_transaction_id), so create_all resolves this one inline.
    matched_expense_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("expenses.id", name="fk_corporate_card_transactions_matched_expense_id"),
        index=True,
    )
    reconciliation_status: Mapped[ReconciliationStatus] = mapped_column(
        Enum(ReconciliationStatus, native_enum=False, length=20),
        default=ReconciliationStatus.unmatched,
        nullable=False,
    )
    import_batch: Mapped[str | None] = mapped_column(String(100))
    raw: Mapped[dict | None] = mapped_column(JSONB)

    organization_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), nullable=False, index=True
    )

    __table_args__ = (
        Index(
            "uq_corporate_card_txn_external",
            "organization_id",
            "external_txn_id",
            unique=True,
            postgresql_where=text("external_txn_id IS NOT NULL"),
        ),
    )


class Expense(Base, EntityMixin, TimestampMixin):
    """A single expense line — out-of-pocket or card-funded."""

    __tablename__ = "expenses"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    # Optional — an expense can exist before being grouped into a report.
    report_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("expense_reports.id"), index=True
    )
    expense_date: Mapped[date] = mapped_column(Date, nullable=False)
    merchant: Mapped[str | None] = mapped_column(String(255))
    category: Mapped[str | None] = mapped_column(String(100))
    description: Mapped[str | None] = mapped_column(Text)

    # --- Money (Numeric, never float) -------------------------------------
    amount: Mapped[Decimal] = mapped_column(Numeric(15, 2), nullable=False)
    currency: Mapped[str] = mapped_column(String(3), default="USD")

    # --- Locked conversion into the OWNING REPORT's currency (issue #157) --
    # A report legitimately mixes currencies (one trip, several countries), so
    # each line carries its own amount plus this rate-locked expression of it in
    # the report's currency. The rate is locked on the write paths that change
    # what needs converting (create-with-report, amount/currency edit, attach,
    # report-currency change) and read back verbatim — a report's total never
    # drifts with the market. NULL when the line is unattached, or (legacy rows)
    # never locked: a foreign-currency line with no lock is counted as
    # *unconverted* and blocks submission rather than summing at face value.
    # See ``services/expense_currency.py``.
    converted_currency: Mapped[str | None] = mapped_column(String(3))
    converted_amount: Mapped[Decimal | None] = mapped_column(Numeric(15, 2))
    converted_fx_rate: Mapped[Decimal | None] = mapped_column(Numeric(18, 8))
    converted_fx_locked_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))

    gl_account_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("gl_accounts.id"), index=True
    )
    receipt_file_key: Mapped[str | None] = mapped_column(String(500))
    payment_method: Mapped[ExpensePaymentMethod] = mapped_column(
        Enum(ExpensePaymentMethod, native_enum=False, length=20),
        default=ExpensePaymentMethod.out_of_pocket,
        nullable=False,
    )
    # Alter side of the expenses↔transactions cycle — use_alter so create_all
    # emits this FK as a deferred ALTER once both tables exist.
    card_transaction_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey(
            "corporate_card_transactions.id",
            use_alter=True,
            name="fk_expenses_card_transaction_id",
        ),
        index=True,
    )
    # List of policy-violation dicts (advisory; no PII).
    policy_violations: Mapped[list | None] = mapped_column(JSONB)
    status: Mapped[ExpenseStatus] = mapped_column(
        Enum(ExpenseStatus, native_enum=False, length=20),
        default=ExpenseStatus.draft,
        nullable=False,
    )
    reimbursable: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    mileage_miles: Mapped[Decimal | None] = mapped_column(Numeric(10, 2))

    organization_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), nullable=False, index=True
    )

    report: Mapped["ExpenseReport | None"] = relationship(back_populates="expenses")


class ExpensePolicy(Base, EntityMixin, TimestampMixin):
    """A reimbursement policy — per-diem, mileage, category limits, pre-approval
    thresholds. Consumed by later workflows (WF3 policy enforcement)."""

    __tablename__ = "expense_policies"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    # NULL = applies to all categories.
    category: Mapped[str | None] = mapped_column(String(100))

    # --- Money / rates (Numeric, never float) -----------------------------
    per_diem_amount: Mapped[Decimal | None] = mapped_column(Numeric(15, 2))
    per_diem_currency: Mapped[str] = mapped_column(String(3), default="USD")
    mileage_rate: Mapped[Decimal | None] = mapped_column(Numeric(10, 4))
    category_limit: Mapped[Decimal | None] = mapped_column(Numeric(15, 2))
    requires_preapproval_above: Mapped[Decimal | None] = mapped_column(Numeric(15, 2))
    requires_receipt_above: Mapped[Decimal | None] = mapped_column(Numeric(15, 2))
    rules: Mapped[dict | None] = mapped_column(JSONB)

    organization_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), nullable=False, index=True
    )


class ExpensePreapproval(Base, EntityMixin, TimestampMixin):
    """A spend pre-approval request raised before an expense is incurred.
    Consumed by later workflows (WF3 pre-approval gating)."""

    __tablename__ = "expense_preapprovals"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    requester_user_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), index=True)
    title: Mapped[str] = mapped_column(String(255), nullable=False)

    # --- Money (Numeric, never float) -------------------------------------
    estimated_amount: Mapped[Decimal] = mapped_column(Numeric(15, 2), nullable=False)
    currency: Mapped[str] = mapped_column(String(3), default="USD")

    category: Mapped[str | None] = mapped_column(String(100))
    justification: Mapped[str | None] = mapped_column(Text)
    status: Mapped[PreapprovalStatus] = mapped_column(
        Enum(PreapprovalStatus, native_enum=False, length=20),
        default=PreapprovalStatus.pending,
        nullable=False,
    )
    decided_by: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True))
    decided_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    expense_report_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("expense_reports.id"), index=True
    )

    organization_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), nullable=False, index=True
    )
